pkgs/development/compilers/ghc/9.2.1.nix

```diff
@@ -107,8 +107,8 @@ let
   '' + lib.optionalString (!enableProfiledLibs) ''
     GhcLibWays = "v dyn"
   '' + lib.optionalString enableRelocatedStaticLibs ''
-    GhcLibHcOpts += -fPIC
-    GhcRtsHcOpts += -fPIC
+    GhcLibHcOpts += -fPIC -fexternal-dynamic-refs
+    GhcRtsHcOpts += -fPIC -fexternal-dynamic-refs
```

Comment:
Would you be able to add a comment here why `-fexternal-dynamic-refs` is needed for `enableRelocatedStaticLibs`?

From the GHC manual: https://downloads.haskell.org/ghc/latest/docs/html/users_guide/phases.html#ghc-flag--fexternal-dynamic-refs

> When generating code, assume that entities imported from a different module might be dynamically linked. This flag is enabled automatically by [-dynamic](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/phases.html#ghc-flag--dynamic).

It is not clear to me why this would be necessary if we are statically linking things.  (Although I don't have a good understanding of static linking, so maybe it should be obvious.)

Reply:
I have to admit I'm baffled by this too and was kind of expecting the haskell maintainers would know better.
Basically, I'm implementing the solution proposed in this blog post: https://www.tweag.io/blog/2020-09-30-bazel-static-haskell/.

Reply:
Thanks for the link to the blog post.  It says:

> When evaluating template Haskell splices, GHC will execute compiled expressions in its built-in bytecode interpreter and this code has to be compatible with the RTS of GHC itself. In short, a GHC configured with a dynamic RTS will not be able to load static Haskell libraries to evaluate template Haskell splices.

It then goes on to say:

> We need to build GHC with a static RTS and to make sure that Haskell code is compiled as position independent code so that it can be loaded into a running GHC for template Haskell splices.

Both of these statements make sense of course, but they don't seem to explain why `-fexternal-dynamic-refs` is necessary.

Reply:
The GHC man page reads:

> Generate code for linking against dynamic libraries

I assume this means that the RTS will have some kind of dynamic library loading facilities even of static (Haskell?) archives, so the byte code interpreter / whatever used by TH will actually work. `-fPIC` also serves this purpose.